pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -74,51 +61,93 @@ def restore_weights(self, model: LightningModule) -> None:
-        fs = get_filesystem(checkpoint_path)
-        if not fs.exists(checkpoint_path):
-            rank_zero_warn("No checkpoint file exists at `resume_from_checkpoint`. Start from scratch")
-            return False
+        Attempt to apply checkpoint states to model/training in this priority:
+        1. from HPC weights
+        2. from `resume_from_checkpoint` file
+        3. don't apply
 
-        # read a checkpoint dictionary object from the 'PyTorch-Lightning checkpoint' file at `checkpoint_path`
-        checkpoint = pl_load(checkpoint_path, map_location=lambda storage, loc: storage)
+        Returns:
+            True if applied else False
+        """
+        # Design Note:
+        #   `attempt_to_restore` has responsibility to whole state restoration flow (e.g. OOM, parallel processing).
+        #   This method has responsibility to applying/assigning state value from nullable checkpoint.
 
-        # acquire the model
-        model = self.trainer.get_model()
+        restored: bool = False
 
-        # restore model and datamodule state
-        self.restore_model_state(model, checkpoint)
+        # 1. Attempt to apply HPC checkpoint.
+        dir_path_hpc = str(self.trainer.weights_save_path)
+        max_suffix = self.max_ckpt_in_folder(dir_path_hpc, "hpc_ckpt_")
+        if max_suffix is not None:
+            checkpoint_path = f'{dir_path_hpc}/hpc_ckpt_{max_suffix}.ckpt'
+            checkpoint = self.restore_states(model, checkpoint_path, self.trainer.root_gpu)
+            model.on_hpc_load(checkpoint)
+            restored = True
+            rank_zero_info(f'restored hpc model from: {checkpoint_path}')
 
-        if on_gpu:
-            model.cuda(self.trainer.root_gpu)
+        # 2. Attempt to apply `resume_from_checkpoint` file.
+        elif self.trainer.resume_from_checkpoint is not None and not self.trainer.testing:
+            adress_checkpoint: str = self.trainer.resume_from_checkpoint
+            if get_filesystem(adress_checkpoint).exists(adress_checkpoint):
+                self.restore_states(model, adress_checkpoint, self.trainer.on_gpu)
+                restored = True
+                rank_zero_info(f"States restored from the checkpoint file at {adress_checkpoint}")
+            else:
+                rank_zero_warn(f"checkpoint file at {adress_checkpoint} does not exist.")
 
-        # restore training state
-        self.restore_training_state(checkpoint)
+        # 3. Do not apply, start from scratch.
+        else:
+            rank_zero_info("Start from scratch.")
 
-        rank_zero_info(f"Restored states from the checkpoint file at {checkpoint_path}")
-        return True
+        return restored
 
-    def restore_model_state(self, model: LightningModule, checkpoint) -> None:
+    def restore_states(
+            self,
+            model: LightningModule,
+            checkpoint_path: str,
+            with_gpu: Union[bool, Optional[int]],
```

Comment:
If what you want is to allow `bool`, `int`, `None` Type, would't `Optional[Union[bool, int]]` be better?

Reply:
It could be one option!  
Current implementation assume `with_gpu=trainer.on_gpu` (**bool**) or `with_gpu=trainer.root_gpu` (**Optional[int]**) as noted in code.  
The method do not assume `Optional[bool]`, so we could work as safety bar.  
But I have no strong opinion about this typing, I am grad to adopt reviews opinion!

https://github.com/PyTorchLightning/pytorch-lightning/blob/7400378a3ead639fd3bc80efd13b1d5104510600/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L118

Reply:
`Optional[x]` expands as `Union[x, None]`, so we can see:
`Union[bool, Optional[int]] = Union[bool, Union[int, None]] = Union[bool, int, None] = Optional[Union[bool, int]]`

Reply:
Yes, this is just an expression. However, if this code wants to allocate `int` and `bool`, and wants `None` in the meaning of default, I think `Optional[Union[int, bool]]` seems the most obvious. :smile: 
But, this isn't essential :smile:

Reply:
Wow, that makes sense!  
I will fix it.  
Thanks for great and interesting suggestion!

Reply:
Fixed.